### PR TITLE
add linear operation and modify one of the tests

### PR DESCRIPTION
**Pull request by migeed-z**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61310
* #61265
* #61239
* **#61238**
* #61188

Differential Revision: [D29626333](https://our.internmc.facebook.com/intern/diff/D29626333)

### CODE REVIEW

File: torch/fx/experimental/graph_gradual_typechecker.py [hunk 1]
```diff
@@ -327,6 +327,56 @@ def maxpool2d_inference_rule(n: Node, op_type):
...
+            new_type_args = list(tensor_type.__args__)
+            new_type_args[-1] = op_type.out_features
+            return TensorType(tuple(new_type_args))
+        else:
+            raise TypeError(f'Inconsistent {op_type.in_features} and {tensor_type.__args__[-1]} in {op_type}')
+    else:
+        raise TypeError(f'Type {tensor_type} must have rank 2 or more.')
+
+@register_inference_rule(torch.nn.Linear)
+def linear_inference_rule(n: Node, op_type):
```
Comment (jamesr66a) [inline torch/fx/experimental/graph_gradual_typechecker.py:348]: Rename `op_type` :)

**Review by jamesr66a — APPROVED** [review]
LGTM

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626333).

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626333).

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626333).

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626333).

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626333).

**Pull request closed without merging**